cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/store/DatasetBasedStreamSizeScheduleStore.java

```diff
@@ -363,25 +395,99 @@ String removeAppVersion(String scheduleId) {
    */
-  public void upgrade()
-    throws InterruptedException, TransactionFailureException, IOException, DatasetManagementException {
-    initialize();
-    factory.createExecutor(ImmutableList.of((TransactionAware) table))
-      .execute(new TransactionExecutor.Subroutine() {
-        @Override
-        public void apply() {
-          upgradeVersionKeys();
+  public void upgrade() throws InterruptedException, IOException, DatasetManagementException {
+    // Wait until the store is initialized
+    while (!storeInitialized.get()) {
+      TimeUnit.SECONDS.sleep(10);
+    }
+
+    final AtomicBoolean upgradeComplete = new AtomicBoolean(false);
+    try {
+      factory.createExecutor(ImmutableList.of((TransactionAware) table))
+        .execute(new TransactionExecutor.Subroutine() {
+          @Override
+          public void apply() throws Exception {
+            byte[] result = table.get(APP_VERSION_UPGRADE_KEY, Bytes.toBytes('c'));
+            if (result != null) {
+              upgradeComplete.set(true);
+            }
+          }
+        });
+    } catch (TransactionFailureException e) {
+      // It is ok to ignore this error since this is an optimization check.
+      String errorMessage = "Initial check to see whether upgrade of StreamSizeScheduleStore is " +
+        "complete or not, has failed. Proceeding with upgrade.";
+      if (e instanceof TransactionConflictException | e instanceof TransactionNotInProgressException) {
+        LOG.trace(errorMessage, e);
+      } else {
+        LOG.error(errorMessage, e);
+      }
+    }
+
+    // If upgrade is already complete, then simply return.
+    if (upgradeComplete.get()) {
+      LOG.info("StreamSizeScheduleStore is already upgraded.");
+      return;
+    }
+
+    final AtomicInteger maxRows = new AtomicInteger(10);
+    final AtomicInteger sleepTimeInMins = new AtomicInteger(1);
+    LOG.info("Starting upgrade of StreamSizeScheduleStore.");
+    while (!isUpgradeComplete()) {
+      sleepTimeInMins.set(1);
+      try {
+        factory.createExecutor(ImmutableList.of((TransactionAware) table))
+          .execute(new TransactionExecutor.Subroutine() {
+            @Override
+            public void apply() {
+              if (upgradeVersionKeys(table, maxRows.get())) {
+                // Upgrade is complete. Mark that app version upgrade is complete in the table.
+                table.put(APP_VERSION_UPGRADE_KEY, Bytes.toBytes('c'),
+                          Bytes.toBytes(ProjectInfo.getVersion().toString()));
+              }
+            }
+          });
+      } catch (TransactionFailureException e) {
+        if (e instanceof TransactionConflictException) {
+          LOG.debug("Upgrade step faced Transaction Conflict exception. Retrying operation now.", e);
+          sleepTimeInMins.set(0);
+        } else if (e instanceof TransactionNotInProgressException) {
+          int currMaxRows = maxRows.get();
+          if (currMaxRows > 1) {
```

Comment:
Just to be safe, let's keep the min to be 500. We don't want to upgrade one row at a time.